backends/bmv2/psa_switch/main.cpp

```diff
@@ -41,37 +43,60 @@ int main(int argc, char *const argv[]) {
     options.langVersion = CompilerOptions::FrontendVersion::P4_16;
     options.compilerVersion = BMV2_PSA_VERSION_STRING;
 
-    if (options.process(argc, argv) != nullptr)
-        options.setInputFile();
+    if (options.process(argc, argv) != nullptr) {
+            if (options.loadIRFromJson == false)
+                    options.setInputFile();
+    }
     if (::errorCount() > 0)
         return 1;
 
     auto hook = options.getDebugHook();
 
     // BMV2 is required for compatibility with the previous compiler.
     options.preprocessor_options += " -D__TARGET_BMV2__";
-    auto program = P4::parseP4File(options);
-    if (program == nullptr || ::errorCount() > 0)
-        return 1;
-    try {
-        P4::P4COptionPragmaParser optionsPragmaParser;
-        program->apply(P4::ApplyOptionsPragmas(optionsPragmaParser));
 
-        P4::FrontEnd frontend;
-        frontend.addDebugHook(hook);
-        program = frontend.run(options, program);
-    } catch (const Util::P4CExceptionBase &bug) {
-        std::cerr << bug.what() << std::endl;
-        return 1;
+    const IR::P4Program *program = nullptr;
+    const IR::ToplevelBlock* toplevel = nullptr;
+
+
+    if (options.loadIRFromJson == false) {
+        program = P4::parseP4File(options);
+
+        if (program == nullptr || ::errorCount() > 0)
+            return 1;
+        try {
+            P4::P4COptionPragmaParser optionsPragmaParser;
+            program->apply(P4::ApplyOptionsPragmas(optionsPragmaParser));
+
+            P4::FrontEnd frontend;
+            frontend.addDebugHook(hook);
+            program = frontend.run(options, program);
+        } catch (const Util::P4CExceptionBase &bug) {
+            std::cerr << bug.what() << std::endl;
+            return 1;
+        }
+        if (program == nullptr || ::errorCount() > 0)
+            return 1;
```

Comment:
It looks like --toJSON dumps after the mid-end, but you are loading BEFORE the mid-end.
Shouldn't you also skip the mid-end when loading json?

Reply:
There should probably be a --toJSON option to dump before the midend (after just the frontend).  Possibly other points make sense as well.  So perhaps a 3rd option that specifies which point in the compile that the json is dumped (for --toJSON) and which point we continue with (for --fromJSON).  But for now, the --toJSON/--fromJSON options should be duals, acting at the same point in the compile.  Whether that point is before the midend or after is IMO not that important, though others may desire a particular spot.  So for minimal change, that should probably be after the midend (for now), as that is what --toJSON currently does

Reply:
Thank you for all comments. I will investigate if we can skip the midEnd, but the issue was generating toplevel without midEnd.process.